model/src/main/kotlin/DependencyGraph.kt

```diff
@@ -0,0 +1,240 @@
+     * on this mapping, the structure with [Scope]s can be generated.
+     */
+    private fun constructReferenceMapping(managerName: String): Map<String, PackageReference> {
+        val refMapping = mutableMapOf<String, PackageReference>()
+        trees.forEach { tree ->
+            constructReferenceTree(managerName, tree, refMapping)
+        }
+
+        return refMapping
+    }
+
+    /**
+     * Construct the tree with [PackageReference]s by navigating the dependency tree starting with [ref] and
+     * populate the given [refMapping]. Use [managerName] to construct package identifiers.
+     */
+    private fun constructReferenceTree(
+        managerName: String,
+        ref: DependencyReference,
+        refMapping: MutableMap<String, PackageReference>
+    ):
+            PackageReference {
+        val indexKey = DependencyIndex.generateKey(ref.index, ref.fragmentIndex)
+        return refMapping.getOrPut(indexKey) {
+            val dependencies = ref.dependencies.mapTo(sortedSetOf()) {
+                constructReferenceTree(managerName, it, refMapping)
+            }
+
+            val id = "${identifierType(ref.dependencyType, managerName)}:${identifierList[ref.index]}"
+            PackageReference(
+                id = Identifier(id), dependencies = dependencies,
+                linkage = ref.linkage, issues = ref.issues
+            )
+        }
+    }
+}
+
+/**
+ * A data class representing the index of a dependency.
+ */
+data class DependencyIndex(
+    /**
+     * The index of the dependency referenced by this object. Each package acting as a dependency is assigned a
+     * unique index. Storing an index rather than an identifier reduces the amount of memory consumed by this
+     * representation.
+     */
+    val index: Int,
+
+    /**
+     * The index of the fragment of the dependency graph this reference points to. This is used to distinguish between
+     * packages that occur multiple times in the dependency graph with different dependencies.
+     */
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    val fragmentIndex: Int = 0
+) {
+    companion object {
+        /**
+         * Generate a string representation for the given [index] and [fragmentIndex] that is used internally as key
+         * in maps.
+         */
+        fun generateKey(index: Int, fragmentIndex: Int): String = "$index.$fragmentIndex"
+    }
+
+    /**
+     * Generate a string key to represent this index.
+     */
+    fun toKey(): String = generateKey(index, fragmentIndex)
+}
+
+/**
+ * An enumeration class defining the types of dependencies.
+ *
+ * This type is used to construct a correct [Identifier] when loading the dependency graph from a result file.
+ */
+enum class DependencyType {
+    /** Indicates that this dependency has been resolved by the current analyzer. */
+    MANAGER,
+
+    /** Indicates that this dependency has been resolved using Maven. */
+    MAVEN,
```

Comment:
Why is Maven special here?

Reply:
The Gradle analyzer implementation uses Maven to resolve packages. It uses three values for the `type` property of generated identifiers:
* Its own manager name ("Gradle") for project dependencies
* "Unknown" for packages without a pom file
* "Maven" for all other packages.

So this is not a change introduced by this PR, but just reflects the status quo. Here is the original code fragment from `Gradle.kt` (line 330ff):

```
      return if (dependency.localPath != null) {
            val id = Identifier(managerName, dependency.groupId, dependency.artifactId, dependency.version)
            PackageReference(id, PackageLinkage.PROJECT_DYNAMIC, transitiveDependencies.toSortedSet(), issues)
        } else {
            val type = dependency.pomFile?.let { "Maven" } ?: "Unknown"
            val id = Identifier(type, dependency.groupId, dependency.artifactId, dependency.version)
            PackageReference(id, dependencies = transitiveDependencies.toSortedSet(), issues = issues)
        }
```

Reply:
This information no longer has to be written to the result file.